src/Form/DateInput.js

```diff
@@ -0,0 +1,74 @@
+import React, { useState, useEffect, useRef } from 'react';
+import PropTypes from 'prop-types';
+import DateFormatter from 'cleave.js/src/shortcuts/DateFormatter';
+import Input from './Input';
+import { createEasyInput } from './EasyInput';
+import { getNextCursorPosition, isDeletingCharacter } from '../utils';
+
+const formatDate = (datePattern, dateString = '') => {
+  const formatter = new DateFormatter(datePattern);
+
+  // Process our date string, bounding values between 1 and 31, and prepending 0s for
+  // for single digit blocks that can't have 2 numbers, e.g. 5
+  let tmpDate = formatter.getValidatedDate(`${dateString}`);
+
+  // Blocks look something like [2, 2, 4], telling us how long each chunk should be
+  return formatter.getBlocks().reduce((str, blockLength, index, blockArr) => {
```

Comment:
`getValidatedDate` returns digits only, so we'll lose pre-existing slashes in the middle of our date string: https://github.com/nosir/cleave.js/blob/master/src/shortcuts/DateFormatter.js#L40

Reply:
I see, could we use splice?
```
var date = dateStr.split("");
date.splice(2,0,"/");
date.splice(5,0,"/");
formattedDate = date.join("");
```

Reply:
I like `getBlocks()` here because we can pass a different date pattern to the component if we wanted to pass a custom pattern and delimiter for, say, an ISO date `2001-04-24`:

```
<DateInput delimiter="-" pattern={['Y', 'm', 'd']} />
```

Reply:
Thanks for explaining the context. Good to know we'd want to be able to support multiple formats. Could it be of interest to do something like this?
```
var dateStr = "1223198"; // should handle any length string appropriately
var reg = new RegExp(`(\\d{${blocks[0]}})(\\d{${blocks[1]}})?(\\d{${blocks[2]}})?`);
var formattedDate = dateStr.replace(/(\d{blocks[0]})(\d{blocks[1]})?(\d{blocks[2]})?/, '$1/$2/$3');
var formattedDate = dateStr.replace(reg, '$1/$2/$3');
// => "12/23/198"
```

Reply:
@cehsu your suggestion outputs `11//` when `dateStr` is `11`. You can pull down the branch, make sure all tests pass with your suggestion, and assert the performance is better by exporting the two formatting methods and adding the following to `DateInput.spec.js`:

```js
const TIMES = 100000;
test('reduce-vs-regex', () => {
  const reduceStart = Date.now();
  for (let i = 0; i < TIMES; i++) {
    formatDate(['m', 'd', 'Y'], '/', '02/05/2000');
  }
  const reduceLength = Date.now() - reduceStart;

  const regexStart = Date.now();
  for (let i = 0; i < TIMES; i++) {
    formatDateRegex(['m', 'd', 'Y'], '/', '02/05/2000');
  }
  const regexLength = Date.now() - regexStart;

  console.log(`Reduce took: ${reduceLength}ms`);
  console.log(`Regex took: ${regexLength}ms`);

  expect(regexLength).toBeLessThan(reduceLength);
});
```

Using a variant of your suggestion that passes our test cases, I'm receiving the following on 100,000 runs:
```diff
+ Reduce: 681ms
- Regex: 755ms
```

Reply:
Thanks @kylealwyn, I didn't check the perf. Indeed, both the splice option (fixed above) and the regex option (fixed above) are slower than the reduce implementation (58%, and 99% respectively).